src/main/java/uk/gov/hmcts/reform/bulkscan/payment/processor/service/servicebus/PaymentMessageProcessor.java

```diff
@@ -0,0 +1,203 @@
+     */
+    public boolean processNextMessage() throws ServiceBusException, InterruptedException {
+        IMessage message = messageReceiver.receive();
+        if (message != null) {
+            MessageProcessingResult result = process(message);
+            tryFinaliseProcessedMessage(message, result);
+        }
+
+        return message != null;
+    }
+
+    private MessageProcessingResult process(IMessage message) {
+        log.info("Started processing message with ID {}", message.getMessageId());
+
+        PaymentMessage payment = null;
+
+        try {
+            payment = paymentMessageParser.parse(message.getMessageBody());
+            logMessageParsed(message, payment);
+            paymentMessageHandler.handlePaymentMessage(payment);
+            log.info("Processed message with ID {}. Envelope ID: {}", message.getMessageId(), payment.envelopeId);
+            return new MessageProcessingResult(SUCCESS);
+        } catch (InvalidMessageException ex) {
+            log.error("Rejected message with ID {}, because it's invalid", message.getMessageId(), ex);
+            return new MessageProcessingResult(UNRECOVERABLE_FAILURE, ex);
+        } catch (PayHubClientException ex) {
+            logMessageProcessingError(message, payment, ex);
+            if (ex.getStatus() == HttpStatus.BAD_REQUEST.CONFLICT) {
+                return new MessageProcessingResult(SUCCESS, ex);
+            }
+        } catch (Exception ex) {
+            logMessageProcessingError(message, payment, ex);
+            return new MessageProcessingResult(POTENTIALLY_RECOVERABLE_FAILURE);
+        }
+
+        return new MessageProcessingResult(POTENTIALLY_RECOVERABLE_FAILURE);
+    }
+
+    private void tryFinaliseProcessedMessage(IMessage message, MessageProcessingResult processingResult) {
+        try {
+            finaliseProcessedMessage(message, processingResult);
+        } catch (InterruptedException ex) {
+            logMessageFinaliseError(message, processingResult.resultType, ex);
+            Thread.currentThread().interrupt();
+        } catch (Exception ex) {
+            logMessageFinaliseError(message, processingResult.resultType, ex);
+        }
+    }
+
+    private void finaliseProcessedMessage(
+        IMessage message,
+        MessageProcessingResult processingResult
+    ) throws InterruptedException, ServiceBusException {
+
+        switch (processingResult.resultType) {
+            case SUCCESS:
+                messageReceiver.complete(message.getLockToken());
+                log.info("Message with ID {} has been completed", message.getMessageId());
+                break;
+            case UNRECOVERABLE_FAILURE:
+                deadLetterTheMessage(
+                    message,
+                    "Message processing error",
+                    processingResult.exception.getMessage()
+                );
+                break;
+            case POTENTIALLY_RECOVERABLE_FAILURE:
+                // starts from 0
```

Comment:
nicer if we can extract code for this case clause into a separate method

Reply:
moved